tests/sync/util.js

```diff
@@ -0,0 +1,194 @@
+    sp15: {
+      assessments: {
+        exam1: {
+          uuid: '1423',
+        },
+      },
+      courseInstance: {
+        uuid: '4321',
+      },
+    },
+  },
+};
+
+/** @typedef {{ assessments: object, courseInstance: object }} CourseInstanceData */
+/** @typedef {{ course: object, questions: { [id: string]: object }, courseInstances: { [id: string]: CourseInstanceData } }} CourseData */
+
+/**
+ * Accepts a CourseData object and creates a PrairieLearn course directory
+ * structure from it. Returns the path to the newly-created directory.
+ * 
+ * @param {CourseData} courseData - The course data to write to disk
+ * @returns string - The path to the directory containing the course data
+ */
+module.exports.writeCourseToTempDirectory = async function(courseData) {
+  const { path: coursePath } = await tmp.dir({ unsafeCleanup: true });
+  await this.writeCourseToDirectory(courseData, coursePath);
+  return coursePath;
+};
+
+/**
+ * Accepts a CourseData object and writes it as a PrairieLearn course
+ * into the given directory. Removes any existing content from the
+ * directory.
+ * 
+ * @param {CourseData} courseData - The course data to write to disk
+ * @param {string} coursePath - The path to the directory to write to
+ */
+module.exports.writeCourseToDirectory = async function(courseData, coursePath) {
+  await fs.emptyDir(coursePath);
+
+  // courseInfo.json
+  const courseInfoPath = path.join(coursePath, 'infoCourse.json');
+  await fs.writeJSON(courseInfoPath, courseData.course);
+
+  // Write all questions
+  const questionsPath = path.join(coursePath, 'questions');
+  await fs.ensureDir(questionsPath);
+  for (const qid of Object.keys(courseData.questions)) {
+    const questionPath = path.join(questionsPath, qid);
+    await fs.ensureDir(questionPath);
+    const questionInfoPath = path.join(questionPath, 'info.json');
+    await fs.writeJSON(questionInfoPath, courseData.questions[qid]);
+  }
+
+  // Write all course instances
+  const courseInstancesPath = path.join(coursePath, 'courseInstances');
+  await fs.ensureDir(courseInstancesPath);
+  for (const shortName of Object.keys(courseData.courseInstances)) {
+    const courseInstance = courseData.courseInstances[shortName];
+    const courseInstancePath = path.join(courseInstancesPath, shortName);
+    await fs.ensureDir(courseInstancePath);
+    const courseInstanceInfoPath = path.join(courseInstancePath, 'infoCourseInstance.json');
+    await fs.writeJSON(courseInstanceInfoPath, courseInstance.courseInstance);
+
+    // Write all assessments for this course instance
+    for (const assessmentName of Object.keys(courseInstance.assessments)) {
+      const assessmentPath = path.join(courseInstancePath, assessmentName);
+      await fs.ensureDir(assessmentPath);
+      const assessmentInfoPath = path.join(assessmentPath, 'infoAssessment.json');
+      await fs.writeJSON(assessmentInfoPath, courseInstance.assessments[assessmentName]);
+    }
+  }
+};
+
+const course = {
```

Comment:
I’m not sure how final this currently is, but I feel like we’d get a more complete test if we had at least two of each object (e.g., two questions, two tags, etc).

Reply:
Still incomplete - thinking about how best to scale this to N of each thing without having to manually repeat the checking code everywhere.

Reply:
It would be awesome to have some way of saying “assert(everything is like this old version, except for this specific set of rows)”. I’m not sure how best to achieve this.

Reply:
I'm hoping some nice generic function like that is just going to kind of naturally fall out of the tests as I write them 🙂 but something something no premature abstractions.